src/main/java/net/pms/dlna/DLNAProtocolInfo.java

```diff
@@ -0,0 +1,265 @@
+	 */
+	public DLNAProtocolInfo(ProtocolInfo template) {
+		this(template.getProtocol(),
+			template.getNetwork(),
+			template.getContentFormat(),
+			template.getAdditionalInfo()
+		);
+	}
+
+	/**
+	 * @return The {@link DLNAOrgFlags} for this {@link DLNAProtocolInfo}.
+	 */
+	public DLNAOrgFlags getFlags() {
+		return flags != null ? flags : DLNAOrgFlags.IMPLIED;
+	}
+
+	/**
+	 * @return The {@code DLNA.ORG_PN} (DLNA media format profile) for this
+	 *         {@link DLNAProtocolInfo} or {@code null} if none is defined.
+	 */
+	public String getPN() {
+		return pn;
+	}
+
+	/**
+	 * Creates a new {@link MimeType} from the content format of this
+	 * {@link DLNAProtocolInfo}.
+	 *
+	 * @return The corresponding {@link MimeType}.
+	 * @throws ParseException If {@code contentFormat} can't be parsed to a
+	 *             {@link MimeType}.
+	 */
+	public MimeType getMimeType() throws ParseException {
+		return MimeType.valueOf(contentFormat);
+	}
+
+	/**
+	 * Creates a new {@link org.seamless.util.MimeType} from the content format
+	 * of this {@link DLNAProtocolInfo}. To create a {@link MimeType}, use
+	 * {@link #getMimeType()} instead.
+	 *
+	 * @return The corresponding {@link org.seamless.util.MimeType}.
+	 * @see #getMimeType()
+	 */
+	@Override
+	public org.seamless.util.MimeType getContentFormatMimeType() throws IllegalArgumentException {
+		return org.seamless.util.MimeType.valueOf(contentFormat);
+	}
+
+	/**
+	 * Parses {@code additionalInfo} and stores the results.
+	 */
+	@Override
+	protected void parseAdditionalInfo() {
+		if (isBlank(additionalInfo)) {
+			return;
+		}
+		String[] parameters = additionalInfo.trim().toUpperCase(Locale.ROOT).split("\\s*;\\s*");
+		for (String parameter : parameters) {
+			if (isBlank(parameter)) {
+				continue;
+			}
+			String[] parameterEntry = parameter.split("\\s*=\\s*");
+			if (parameterEntry.length == 2) {
+				DLNAAttribute.Type type =
+					DLNAAttribute.Type.valueOfAttributeName(parameterEntry[0]);
+				if (type != null) {
+					DLNAAttribute<?> dlnaAttrinute;
+					try {
+						dlnaAttrinute = DLNAAttribute.newInstance(type, parameterEntry[1], getMimeType().toStringWithoutParameters());
+					} catch (ParseException e) {
+						dlnaAttrinute = DLNAAttribute.newInstance(type, parameterEntry[1], null);
+						LOGGER.debug("Error parsing mime-type while parsing additionalInfo from \"{}\": {}", this, e.getMessage());
+						LOGGER.trace("", e);
+					}
+					if (dlnaAttrinute != null) {
+						attributes.put(type, dlnaAttrinute);
+					}
+					switch (type) {
+						case DLNA_ORG_FLAGS:
+							flags = new DLNAOrgFlags(parameterEntry[1]);
+							break;
+						case DLNA_ORG_PN:
+							pn = parameterEntry[1];
+							break;
+						case DLNA_ORG_CI:
+						case DLNA_ORG_OP:
+						case DLNA_ORG_PS:
+						default:
+							break;
+					}
+				} else if (!"SONY.COM_PN".equals(parameterEntry[0])) {
+					LOGGER.debug("Unknown ProtocolInfo parameter type \"{}\" ({})", parameterEntry[0], parameter);
```

Comment:
@Sami32 It's a "standard" operation, it's just that it isn't precise. Java is bad with some such very "basic" things, but the problem is really "deeper" than Java. I don't think I'll be able to explain it in a way that you understand - I don't even know if I can describe it at all in English. But it has to do with how computers store and handle numbers in general. The "standard" integer types (byte, short, int, long) "roll over" when they overflow and the bit that's outside the range is simply "discarded". A byte value of 255 (11111111b) will if you add 1 be 0. This is because when you add 1 to 11111111b it would become 100000000b - which is 9 bits. Since a byte is restricted to 8 bits, the 9'th bit is discarded and all you have left is 00000000b = 0. The same happens for the wider types, just at bigger numbers. When you do power of, the number can quickly become large and overflow occur so that the answer is actually wrong. Java's "solution" to the problem is to not provide such a function and force us to use ```double``` instead. [Floating point](https://en.wikipedia.org/wiki/IEEE_floating_point) types like ```double``` doesn't "roll over" but becomes "infinite" instead, which lets you detect that the result was to big. Floating points can also hold much bigger values, since they're basicly using "scientific notation" by being represented by a value and its exponent. That means that storing very big numbers is "cheap" when it comes to consumed bits, but the downside is that everything that can't be kept in the given bit depth is simply rounded and discarded. ```double``` is also stored in base 2, not base 10, which means that a lot of "simple" values in base 10, like "0.1", can't be stored accurately. This has nothing to do with Java, the basic number types are implemented in the CPU's and is common to all programming languages.

More precise data types like ```BigInteger``` and ```BigDecimal``` exists, but they are much slower to use (since they aren't "backed" by the underlying hardware) and require more memory to store. In many cases that wouldn't actually matter as the extra time and memory consumed would be so tiny, but using them is much more cumbersome so it often isn't very tempting. You can't just type in their values or use the standard ```+```, ```-```, ```*``` etc. operators on them, so it's some extra hazzle involved. I think old habit also plays a role, and the fact that so few methods actually accepts them. In any case, most of this isn't Java's "fault", rounding errors is a "fact of life" with computers in general.

I now understand why I observed what I thought was a very strange behaviour on the Commodore 64 - I noticed that ```7^2``` was ```49.0000001``` (I don't remember exactly how many decimals it were) according to the C64, while ```7*7``` was ```49```. I wondered how they could make such a "fundamental" bug, but I now realize that it probably was because they used ```float``` (or possibly a 16 bit floating point type) for "power of" - much like Java does.

A precision of 1/1000 isn't enough when it comes to play speed, remember that it's often millions or billions of bytes in a video file. Small rounding errors quickly gets large when multiplied with big numbers, which is probably why the DLNA standard explicitly states that this can NOT be stored as a floating point number/be affected by rounding.

When it comes to the ```Rational``` implementation, I don't think like that. I don't try to anticipate how I (or somebody else) will use something in the future, I simply try to make it as robust as possible. That way, it can be used for something I didn't even know "existed" when I wrote it some time in the future. An implementation of something as basic as ```Rational``` isn't limited to UMS, I want to make it once - and use it whenever I need it in the future, for any project. I know I'm a bit mad, but that's the way I think - I want to create "building blocks" that I can use for other things later and that I can trust to behave correctly instead of having a lot of implicit limitations from their original scope.